### PR TITLE
Build run-time image with jlink

### PR DESCRIPTION
**Pull request by plata**
I keep getting `IllegalArgumentException` in the maven build. Maybe anyone else got an idea what I'm doing wrong.

### CODE REVIEW

**Comment by plata** [comment]
Issue is probably: https://issues.apache.org/jira/projects/MJLINK/issues/MJLINK-9?filter=allopenissues

File: phoenicis-dist/pom.xml [hunk 1]
```diff
@@ -145,4 +146,24 @@
             </plugin>
         </plugins>
     </build>
+    <profiles>
+        <profile>
+            <id>jlink-packaging</id>
+            <activation>
+                <jdk>[9</jdk>
```
Comment (qparis) [inline phoenicis-dist/pom.xml]: Is it [9 or 9?

Comment (plata) [reply phoenicis-dist/pom.xml]: `[9` should be correct (9 and greater).

**Comment by madoar** [comment]
I have no prior experience with jlink.
Are you sure, that the dependency to `maven-jlink-plugin` is in the correct `pom.xml` file?
I mean as far as I understand the explanation in https://maven.apache.org/plugins/maven-jlink-plugin/usage.html it could also be correct to define/use it in the dist project.

**Comment by madoar** [comment]
I've just found the following example: https://github.com/khmarbaise/jdk9-jlink-jmod-example/tree/master/maven-example
It not only depends on jlink but also on jmod.

**Comment by madoar** [comment]
Ok, when taking a closed look at the previous mentioned repository, I discovered that he has multiple examples, some of which don't use jmod: https://github.com/khmarbaise/jdk9-jlink-jmod-example

**Comment by plata** [comment]
@madoar I've also tried the dependency in phoenicis-dist already, no difference. I know the examples you've linked but I don't see what I'm doing different. Can you try it?

**Comment by madoar** [comment]
Can you add an `-X` to the mvn clean test command to travis?
It would be interesting to see where exactly the exception is thrown

**Comment by plata** [comment]
I tried that but I didn't really the the issue there either.

**Comment by madoar** [comment]
I can't find a commit in this PR adding the `-X` to the test command used by travis?

**Comment by plata** [comment]
I only did that locally.

**Comment by madoar** [comment]
You need to do it online, if you want travis to execute it.
To be more precise, you need to do it when building/testing the project with Java 11, where the problem occurs.

**Comment by plata** [comment]
I know. What I mean is: I've added `-X` locally and didn't get any useful information from it (or at least wasn't able to understand it properly). Maybe try yourself if you get any insight from it.

**Comment by madoar** [comment]
Ok, I've taken a look at the code of the plugin.
- it looks like the code version used in the plugin doesn't reflect the current codebase on github (https://github.com/apache/maven-jlink-plugin/blob/master/src/main/java/org/apache/maven/plugins/jlink/JLinkMojo.java)
- have you checked, that maven uses the correct java version?
Otherwise, I'm also not sure what the reason for the error might be.
One possibility is, that the dependencies need to be Java 9 compatible modules, but I'm not sure.

**Comment by plata** [comment]
Java version is ok. From my understand, the dependencies can me either modules or just jars (like in our case).

**Comment by plata** [comment]
How would I figure out which asm version shall be used?

**Comment by madoar** [comment]
It seems like the plugin implicitly depends on asm version `6.0_BETA` (maven-jlink-plugin -> plexus-java -> asm).
This version was released on 16 July 2017.
The version logs (https://asm.ow2.io/versions.html) indicate, that version `6.1` introduces a constant for Java 10, which I guess means, that only version 6.1+ is compatible with Java 10.

**Comment by plata** [comment]
Not sure if we still need this. Building a flatpak is possible now, so jlink is not needed for Linux. Do we want it for OSX?

**Comment by madoar** [comment]
I like the jlink solution, because it provides a native installer, providing the necessary JRE version.
This should make us independent of the packages provided by the operating system.

**Comment by plata** [comment]
Flatpak achieves the same thing.

**Comment by madoar** [comment]
Yes, but only on operating systems providing flatpak.
I'm not sure if we should force our users to install flatpak.
Flatpak in my understanding should be one of many available installation mechanism.

**Comment by plata** [comment]
Which linux distribution does not support flatpak? To be honest, I would support only a minimal set of packaging (I would also drop .deb support). It's only a lot of work to maintain and in the end it doesn't help anybody.

**Comment by madoar** [comment]
I don't know which distributions support or don't support flatpak.
It's just that flatpak is a dependency, which the user would require to install our application.
This in my opinion is a bad design, because the user would have no alternatives.

I agree with you that we should provide a limited set of packaging choices, the question is which packaging choices do we provide?
Jlink has the advantage, that it represents a solution, which can be used to generate installers for both osx and linux (as far as I understand it).

**Comment by plata** [comment]
I wouldn't really consider flatpak as a dependency. That's a bit like calling apt a dependency. Also, with deb or rpm we have a lot more dependencies (e.g. system JRE).

Apart from this, we already had a lot of trouble with different runtimes (i.e. Wine or certain apps have runtime dependencies to some libs). With flatpak, we can handle this much better. This is not possible with jlink.

**Comment by madoar** [comment]
About the cause of the errors in this PR:
I think we can override the used asm version by the plugin: https://stackoverflow.com/questions/17422013/override-dependencies-in-a-maven-plugin-one-plugin-different-modules-and-diffe

**Comment by plata** [comment]
I really have no idea how you would normally package applications for OSX. If we can get jlink running, I'm fine using it.

**Comment by madoar** [comment]
I believe the current error means, that the `--add-modules` and `--add-exports` parameters are missing.

**Comment by plata** [comment]
Where should those be added?

**Comment by madoar** [comment]
Right to the configuration of the `maven-jlink-plugin`, see also https://github.com/apache/maven-jlink-plugin/blob/master/src/main/java/org/apache/maven/plugins/jlink/JLinkMojo.java#L144

**Comment by madoar** [comment]
The PR indicates another unrelated problem:
```
[INFO] The following dependencies will be linked into the runtime image:
[INFO]  -> module: phoenicis.javafx ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-javafx/target/phoenicis-javafx-5.0-SNAPSHOT.jar )
[INFO]  -> module: phoenicis.configuration ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-configuration/target/phoenicis-configuration-5.0-SNAPSHOT.jar )
[INFO]  -> module: jgettext ( /home/travis/.m2/repository/org/fedorahosted/tennera/jgettext/0.15.1/jgettext-0.15.1.jar )
[INFO]  -> module: antlr ( /home/travis/.m2/repository/antlr/antlr/2.7.7/antlr-2.7.7.jar )
[INFO]  -> module: com.google.common ( /home/travis/.m2/repository/com/google/guava/guava/26.0-jre/guava-26.0-jre.jar )
[INFO]  -> module: jsr305 ( /home/travis/.m2/repository/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.jar )
[INFO]  -> module: checker.qual ( /home/travis/.m2/repository/org/checkerframework/checker-qual/2.5.2/checker-qual-2.5.2.jar )
[INFO]  -> module: error.prone.annotations ( /home/travis/.m2/repository/com/google/errorprone/error_prone_annotations/2.1.3/error_prone_annotations-2.1.3.jar )
[INFO]  -> module: j2objc.annotations ( /home/travis/.m2/repository/com/google/j2objc/j2objc-annotations/1.1/j2objc-annotations-1.1.jar )
[INFO]  -> module: animal.sniffer.annotations ( /home/travis/.m2/repository/org/codehaus/mojo/animal-sniffer-annotations/1.14/animal-sniffer-annotations-1.14.jar )
[INFO]  -> module: com.fasterxml.jackson.databind ( /home/travis/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.9.7/jackson-databind-2.9.7.jar )
[INFO]  -> module: jackson.annotations ( /home/travis/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.9.0/jackson-annotations-2.9.0.jar )
[INFO]  -> module: com.fasterxml.jackson.core ( /home/travis/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.9.7/jackson-core-2.9.7.jar )
[INFO]  -> module: gettext.commons ( /home/travis/.m2/repository/com/googlecode/gettext-commons/gettext-commons/0.9.8/gettext-commons-0.9.8.jar )
[INFO]  -> module: phoenicis.repository ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-repository/target/phoenicis-repository-5.0-SNAPSHOT.jar )
[INFO]  -> module: phoenicis.entities ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-entities/target/phoenicis-entities-5.0-SNAPSHOT.jar )
[INFO]  -> module: phoenicis.multithreading ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-multithreading/target/phoenicis-multithreading-5.0-SNAPSHOT.jar )
[INFO]  -> module: java.annotation ( /home/travis/.m2/repository/javax/annotation/javax.annotation-api/1.3.2/javax.annotation-api-1.3.2.jar )
[INFO]  -> module: phoenicis.tools ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-tools/target/phoenicis-tools-5.0-SNAPSHOT.jar )
[INFO]  -> module: java.activation ( /home/travis/.m2/repository/com/sun/activation/javax.activation/1.2.0/javax.activation-1.2.0.jar )
[INFO]  -> module: phoenicis.win32 ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-win32/target/phoenicis-win32-5.0-SNAPSHOT.jar )
[INFO]  -> module: org.apache.commons.codec ( /home/travis/.m2/repository/commons-codec/commons-codec/1.11/commons-codec-1.11.jar )
[INFO]  -> module: jmimemagic ( /home/travis/.m2/repository/jmimemagic/jmimemagic/0.1.2/jmimemagic-0.1.2.jar )
[INFO]  -> module: oro ( /home/travis/.m2/repository/oro/oro/2.0.8/oro-2.0.8.jar )
[INFO]  -> module: commons.logging ( /home/travis/.m2/repository/commons-logging/commons-logging/1.0.4/commons-logging-1.0.4.jar )
[INFO]  -> module: xercesImpl ( /home/travis/.m2/repository/xerces/xercesImpl/2.7.1/xercesImpl-2.7.1.jar )
[INFO]  -> module: xml.apis ( /home/travis/.m2/repository/xml-apis/xml-apis/1.0.b2/xml-apis-1.0.b2.jar )
[INFO]  -> module: xmlParserAPIs ( /home/travis/.m2/repository/xml-apis/xmlParserAPIs/2.0.2/xmlParserAPIs-2.0.2.jar )
[INFO]  -> module: org.apache.commons.compress ( /home/travis/.m2/repository/org/apache/commons/commons-compress/1.18/commons-compress-1.18.jar )
[INFO]  -> module: bcpg.jdk16 ( /home/travis/.m2/repository/org/bouncycastle/bcpg-jdk16/1.46/bcpg-jdk16-1.46.jar )
[INFO]  -> module: bcprov.jdk16 ( /home/travis/.m2/repository/org/bouncycastle/bcprov-jdk16/1.46/bcprov-jdk16-1.46.jar )
[INFO]  -> module: bcmail.jdk16 ( /home/travis/.m2/repository/org/bouncycastle/bcmail-jdk16/1.46/bcmail-jdk16-1.46.jar )
[INFO]  -> module: commons.lang ( /home/travis/.m2/repository/commons-lang/commons-lang/2.6/commons-lang-2.6.jar )
[INFO]  -> module: org.apache.commons.io ( /home/travis/.m2/repository/commons-io/commons-io/2.6/commons-io-2.6.jar )
[INFO]  -> module: org.eclipse.jgit ( /home/travis/.m2/repository/org/eclipse/jgit/org.eclipse.jgit/5.1.1.201809181055-r/org.eclipse.jgit-5.1.1.201809181055-r.jar )
[INFO]  -> module: jsch ( /home/travis/.m2/repository/com/jcraft/jsch/0.1.54/jsch-0.1.54.jar )
[INFO]  -> module: jzlib ( /home/travis/.m2/repository/com/jcraft/jzlib/1.1.1/jzlib-1.1.1.jar )
[INFO]  -> module: JavaEWAH ( /home/travis/.m2/repository/com/googlecode/javaewah/JavaEWAH/1.1.6/JavaEWAH-1.1.6.jar )
[INFO]  -> module: phoenicis.scripts ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-scripts/target/phoenicis-scripts-5.0-SNAPSHOT.jar )
[INFO]  -> module: phoenicis.library ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-library/target/phoenicis-library-5.0-SNAPSHOT.jar )
[INFO]  -> module: phoenicis.engines ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-engines/target/phoenicis-engines-5.0-SNAPSHOT.jar )
[INFO]  -> module: phoenicis.containers ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-containers/target/phoenicis-containers-5.0-SNAPSHOT.jar )
[INFO]  -> module: slf4j.api ( /home/travis/.m2/repository/org/slf4j/slf4j-api/1.7.25/slf4j-api-1.7.25.jar )
[INFO]  -> module: logback.classic ( /home/travis/.m2/repository/ch/qos/logback/logback-classic/1.2.3/logback-classic-1.2.3.jar )
[INFO]  -> module: logback.core ( /home/travis/.m2/repository/ch/qos/logback/logback-core/1.2.3/logback-core-1.2.3.jar )
[INFO]  -> module: spring.context ( /home/travis/.m2/repository/org/springframework/spring-context/5.1.0.RELEASE/spring-context-5.1.0.RELEASE.jar )
[INFO]  -> module: spring.aop ( /home/travis/.m2/repository/org/springframework/spring-aop/5.1.0.RELEASE/spring-aop-5.1.0.RELEASE.jar )
[INFO]  -> module: spring.beans ( /home/travis/.m2/repository/org/springframework/spring-beans/5.1.0.RELEASE/spring-beans-5.1.0.RELEASE.jar )
[INFO]  -> module: spring.expression ( /home/travis/.m2/repository/org/springframework/spring-expression/5.1.0.RELEASE/spring-expression-5.1.0.RELEASE.jar )
[INFO]  -> module: spring.core ( /home/travis/.m2/repository/org/springframework/spring-core/5.1.0.RELEASE/spring-core-5.1.0.RELEASE.jar )
[INFO]  -> module: spring.jcl ( /home/travis/.m2/repository/org/springframework/spring-jcl/5.1.0.RELEASE/spring-jcl-5.1.0.RELEASE.jar )
[INFO]  -> module: phoenicis.settings ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-settings/target/phoenicis-settings-5.0-SNAPSHOT.jar )
[INFO]  -> module: fuzzywuzzy ( /home/travis/.m2/repository/me/xdrop/fuzzywuzzy/1.1.10/fuzzywuzzy-1.1.10.jar )
[INFO]  -> module: javafx.controlsEmpty ( /home/travis/.m2/repository/org/openjfx/javafx-controls/11/javafx-controls-11.jar )
[INFO]  -> module: javafx.controls ( /home/travis/.m2/repository/org/openjfx/javafx-controls/11/javafx-controls-11-linux.jar )
[INFO]  -> module: javafx.graphicsEmpty ( /home/travis/.m2/repository/org/openjfx/javafx-graphics/11/javafx-graphics-11.jar )
[INFO]  -> module: javafx.graphics ( /home/travis/.m2/repository/org/openjfx/javafx-graphics/11/javafx-graphics-11-linux.jar )
[INFO]  -> module: javafx.baseEmpty ( /home/travis/.m2/repository/org/openjfx/javafx-base/11/javafx-base-11.jar )
[INFO]  -> module: javafx.base ( /home/travis/.m2/repository/org/openjfx/javafx-base/11/javafx-base-11-linux.jar )
[INFO]  -> module: javafx.webEmpty ( /home/travis/.m2/repository/org/openjfx/javafx-web/11/javafx-web-11.jar )
[INFO]  -> module: javafx.web ( /home/travis/.m2/repository/org/openjfx/javafx-web/11/javafx-web-11-linux.jar )
[INFO]  -> module: javafx.mediaEmpty ( /home/travis/.m2/repository/org/openjfx/javafx-media/11/javafx-media-11.jar )
[INFO]  -> module: javafx.media ( /home/travis/.m2/repository/org/openjfx/javafx-media/11/javafx-media-11-linux.jar )
[INFO]  -> module: phoenicis.cli ( /home/travis/build/PhoenicisOrg/phoenicis/phoenicis-cli/target/phoenicis-cli-5.0-SNAPSHOT.jar )
[INFO]  -> module: commandline ( /home/travis/.m2/repository/com/github/jankroken/commandline/1.7.0/commandline-1.7.0.jar )
[INFO]  -> module: org.jsoup ( /home/travis/.m2/repository/org/jsoup/jsoup/1.11.3/jsoup-1.11.3.jar )
```
When looking at the modules for javafx, you see that only the linux dependencies are in there.
I believe, that we're unable to build a version for osx from a linux system.

**Comment by plata** [comment]
The `addModules` is clear. It was rather about the "add-exports". To the compiler plugin?

I think that's normal. As the JRE is bundled, it has to do it vs your native system. This is the same for all native installations.

**Comment by plata** [comment]
As the flatpak is working, we do not need this anymore.

**Pull request closed without merging**